Mathlib/Algebra/Group/AddChar.lean

```diff
@@ -31,170 +30,236 @@ see `Mathlib.NumberTheory.LegendreSymbol.AddCharacter`.
+/-- Composing a `MonoidHom` with an `AddChar` yields another `AddChar`. -/
+def monoidHom_comp {N : Type*} [Monoid N] (f : M →* N) (φ : AddChar A M) : AddChar A N :=
+  (toMonoidHomEquiv A N).symm (f.comp φ.toMonoidHom)
 
-/-- An additive character maps `0` to `1`. -/
-@[simp]
-theorem map_zero_one (ψ : AddChar A M) : ψ 0 = 1 := by rw [coe_to_fun_apply, ofAdd_zero, map_one]
-#align add_char.map_zero_one AddChar.map_zero_one
+@[simp] lemma coe_monoidHom_comp {N : Type*} [Monoid N] (f : M →* N) (φ : AddChar A M) :
+    monoidHom_comp f φ = f ∘ φ := rfl
 
-/-- An additive character maps sums to products. -/
-@[simp]
-theorem map_add_mul (ψ : AddChar A M) (x y : A) : ψ (x + y) = ψ x * ψ y := by
-  rw [coe_to_fun_apply, coe_to_fun_apply _ x, coe_to_fun_apply _ y, ofAdd_add, map_mul]
-#align add_char.map_add_mul AddChar.map_add_mul
+/-- Composing an `AddChar` with an `AddMonoidHom` yields another `AddChar`. -/
+def comp_addMonoidHom {B : Type*} [AddMonoid B] (φ : AddChar B M) (f : A →+ B) : AddChar A M :=
+  (toAddMonoidHomEquiv A M).symm (φ.toAddMonoidHom.comp f)
 
-/-- An additive character maps multiples by natural numbers to powers. -/
-@[simp]
-theorem map_nsmul_pow (ψ : AddChar A M) (n : ℕ) (x : A) : ψ (n • x) = ψ x ^ n := by
-  rw [coe_to_fun_apply, coe_to_fun_apply _ x, ofAdd_nsmul, map_pow]
-#align add_char.map_nsmul_pow AddChar.map_nsmul_pow
+@[simp] lemma coe_comp_addMonoidHom {B : Type*} [AddMonoid B] (φ : AddChar B M) (f : A →+ B) :
+    φ.comp_addMonoidHom f = φ ∘ f := rfl
 
-end CoeToFun
+/-- An additive character is *nontrivial* if it takes a value `≠ 1`. -/
+def IsNontrivial (ψ : AddChar A M) : Prop := ∃ a : A, ψ a ≠ 1
+#align add_char.is_nontrivial AddChar.IsNontrivial
 
-section GroupStructure
+/-- An additive character is nontrivial iff it is not the trivial character. -/
+lemma isNontrivial_iff_ne_trivial (ψ : AddChar A M) : IsNontrivial ψ ↔ ψ ≠ 1 :=
+  not_forall.symm.trans (DFunLike.ext_iff (f := ψ) (g := 1)).symm.not
 
-open Multiplicative
+#align add_char.is_nontrivial_iff_ne_trivial AddChar.isNontrivial_iff_ne_trivial
 
-variable {A : Type u} [AddCommGroup A] {M : Type v} [CommMonoid M]
+end Basic
 
-/-- An additive character on a commutative additive group has an inverse.
+section toCommMonoid
 
-Note that this is a different inverse to the one provided by `MonoidHom.inv`,
-as it acts on the domain instead of the codomain. -/
-instance hasInv : Inv (AddChar A M) :=
-  ⟨fun ψ => ψ.comp invMonoidHom⟩
-#align add_char.has_inv AddChar.hasInv
+variable {A M : Type*} [AddMonoid A] [CommMonoid M]
 
-theorem inv_apply (ψ : AddChar A M) (x : A) : ψ⁻¹ x = ψ (-x) :=
-  rfl
+/-- When `M` is commutative, `AddChar A M` is a commutative monoid. -/
+instance instCommMonoid : CommMonoid (AddChar A M) := (toMonoidHomEquiv A M).commMonoid
+
+-- Porting note: added
+@[simp] lemma mul_apply (ψ φ : AddChar A M) (a : A) : (ψ * φ) a = ψ a * φ a := rfl
+
+@[simp] lemma pow_apply (ψ : AddChar A M) (n : ℕ) (a : A) : (ψ ^ n) a = (ψ a) ^ n := rfl
```

Comment:
Again for this one, I am not sure whether it should be a `simp`.

Reply:
The LHS involves the monoid structure on `AddChar A M`, the RHS only the monoid structure on `M`, so I think it is fair to describe the RHS being simpler.

Reply:
I see what you mean, but I wonder (and really have no strong opinion) if for the user seeing all the LHS become the RHS is what is hoped for. Of course one can always revert or use a `simp only` to avoid some of these to fire, but to me both sides seem of potentially equal utility.

Reply:
I'm thinking of this as a modified version of things like `Pi.add_apply` etc, which are all `simp`s.